ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java

```diff
@@ -486,10 +486,10 @@ private TransactionWithMetadata transactionByHeaderAndIndex(
    */
   public List<LogWithMetadata> matchingLogs(
       final long fromBlockNumber, final long toBlockNumber, final LogsQuery query) {
-    return LongStream.rangeClosed(fromBlockNumber, toBlockNumber)
-        .mapToObj(blockchain::getBlockHashByNumber)
-        .takeWhile(Optional::isPresent)
-        .flatMap(Optional::stream)
+    return LongStream.rangeClosed(fromBlockNumber, Math.min(toBlockNumber, headBlockNumber()))
+        .mapToObj(blockchain::getBlockHeader)
+        .filter(header -> header.isPresent() && query.couldMatch(header.get().getLogsBloom()))
+        .map(header -> header.get().getHash())
```

Comment:
Interesting interaction here, if we use a `takeWhile` instead of the `Math.min`, we can actually return more logs in the case where a new block was added while the processing of the rpc call is happening because Java Streams are pull based. I lean towards the `takeWhile` approach because if the user asks for a block that's slightly too high, this could make the difference between them seeing their everything they wanted and most of what they wanted (in the case where a block is stored in time for the end of the stream to be reached).

Reply:
That's the sort of subtlety that warrants a comment in the code.  We need to be better about commenting our non-obvious intents.

Reply:
I agree

Reply:
Will keep that in mind for next time

Reply:
With that in mind, I realized it would be better to switch to `.map(Optional::get` instead of `.flatMap(Optional::stream` because the former would complain loudly and obviously if the invariant of all the headers being present is invalidated.